Sources/NIOPosix/SocketChannel.swift

```diff
@@ -672,6 +684,14 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
     override func bufferPendingWrite(data: NIOAny, promise: EventLoopPromise<Void>?) {
         let data = self.unwrapData(data, as: AddressedEnvelope<ByteBuffer>.self)
 
+        // If the datagram socket is connected then we check here that the remote is appropriate.
+        if let remoteAddress = self.remoteAddress {
+            guard data.remoteAddress == remoteAddress else {
+                promise?.fail(IOError(errnoCode: EISCONN, reason: "Socket was connected before flushing pending write."))
+                return
+            }
+        }
```

Comment:
@Lukasa I'm thinking that we might get even more perf win here once PDWM supports `ByteBuffer`. In this branch we could buffer only the buffer, without the envelope. Then we wouldn't even need to do the address stripping on the way out in PDWM flush?

Reply:
👍 Is that something we want in the first PR or can we do this incrementally?

Reply:
OK, great. I added support for sending unaddressed `ByteBuffer` and have added tests showing how users should only use `AddressedEnvelope` with unconnected channels and `ByteBuffer` with connected channels.